<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Jacob WIlliams" />

<meta name="date" content="2024-10-16" />

<title>RICE</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">RICE</h1>
<h4 class="author">Jacob WIlliams</h4>
<h4 class="date">2024-10-16</h4>



<p>RICE performs risk prediction for complex traits and diseases by
incorporating both common (RICE-CV) and rare variants (RICE-RV). RICE
framework consists of three main steps: (1) RICE-CV combines multiple
existing common variant PRSs using ensemble learning to generate a
single robust PRS; (2) RICE-RV identifies significant rare variants sets
conditioned on RICE-CV, models them using statistical methods, and
combines the results through ensemble learning to produce a rare variant
PRS; (3) The PRS from RICE-CV and RICE-RV are then evaluated together in
a regression model, adjusting for covariates. RICE requires three
independent datasets: (1) a training dataset to generate GWAS summary
statistics, rare variant association p-values, and train model for both
common and rare variant models; (2) a tuning dataset to optimize model
parameters and train the ensemble learning step; (3) a validation
dataset to assess the final prediction performance.</p>
<p>This vignette outlines how to construct a rare variant PRS and
provides details on the SuperLearner ensemble regression algorithm used
in both RICE-RV and RICE-CV.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="do">## Create Train/Tune/Validation Partitions</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1330</span>)</span></code></pre></div>
<div id="rice-rv" class="section level1">
<h1>RICE-RV</h1>
<p>RICE-RV performs rare variant association analysis to identify
significant rare variant sets using the training dataset with
STAARpipeline. With these significant rare variant sets, burden scores
are constructed for each set. Then penalized regression models are
trained using the burden scores and predicted onto the tuning and
validation datasets. Ensemble regression using SuperLearner optimizes
the predictions to form a single PRS. Lastly, this PRS is projected onto
the validation set to assess the final prediction performance.</p>
<p>The first step of RICE-RV involves STAARpipeline. RICE-RV can be seen
as an extension of the STAARpipeline, as it relies on the results of
STAARpipeline and operates with identical data structure. Therefore,
before attempting to implement RICE-RV, we suggest reading through the
STAARpipeline tutorial (<a href="https://github.com/xihaoli/STAARpipeline-Tutorial" class="uri">https://github.com/xihaoli/STAARpipeline-Tutorial</a>). This
vignette will only provide brief summaries of conducting rare variant
analysis with STAARpipeline.</p>
<div id="agds-files" class="section level2">
<h2>aGDS Files</h2>
<p>The first step of both RICE-RV is creating annotated genomic data
structure files (aGDS) for each chromosome. In the manuscript, aGDS
files included the train, tune, and validation sets, as we can extract
each set by subsetting. These are highly compressed files that include
all variants and the functional annotations of each variant. These are
commonly created using VCF files. Full details of creating aGDS files
can be found in the STAARpipeline tutorial (<a href="https://github.com/xihaoli/STAARpipeline-Tutorial?tab=readme-ov-file#generate-genomic-data-structure-gds-file" class="uri">https://github.com/xihaoli/STAARpipeline-Tutorial?tab=readme-ov-file#generate-genomic-data-structure-gds-file</a>).</p>
</div>
<div id="baseline-model" class="section level2">
<h2>Baseline Model</h2>
<p>The second step of both RICE-RV is creating a baseline null model to
conduct rare variant association analysis with the training dataset.
RICE-RV is conducted after RICE-CV therefore the baseline model for rare
variant association analysis includes the common variant PRS from
RICE-CV. In the manuscript, RICE-CV was constructed with CT, LDpred2,
and Lassosum2 in the case of the training dataset consisting of a single
ancestry and CT-SLEB, PROSPER, and JointPRS in the case of the training
dataset consisting of multiple ancestries. Full details on the
implementation of this can be found at … and <a href="https://github.com/jwilliams10/RareVariantPRS" class="uri">https://github.com/jwilliams10/RareVariantPRS</a>. However,
in practice RICE-RV can be constructed conditional on any common variant
PRS. Therefore, here we fit the baseline model with a simulated common
variant PRS.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="do">### Fit Baseline Model for Training Data</span></span></code></pre></div>
</div>
<div id="rare-variant-association-study" class="section level2">
<h2>Rare Variant Association Study</h2>
<p>The third step of RICE-RV is to perform rare variant association
analysis using STAARpipeline. Rare variant association analysis within
STAARpipeline can be divided into two analyses; gene-centric coding
analysis and gene-centric noncoding analysis. Gene-centric coding
analysis conducts association analysis with rare variants located in
protein-coding genes in the coding region and gene-centric noncoding
analysis in the noncoding region. Whole exome sequencing data is largely
comprised of rare variants located in the coding region while whole
genome sequencing data has rare variants in both the coding and
noncoding regions. This vignette only provides details for an example
gene-centric coding analysis. Full details on how to implement
gene-centric noncoding analysis and implementing both coding and
noncoding analysis within a HPC is available at <a href="https://github.com/xihaoli/STAARpipeline-Tutorial?tab=readme-ov-file#step-31-gene-centric-coding-analysis" class="uri">https://github.com/xihaoli/STAARpipeline-Tutorial?tab=readme-ov-file#step-31-gene-centric-coding-analysis</a>.</p>
<p>Here we implement gene-centric coding analysis on the training
dataset. Note this is a non-optimal way of implementing this. If you
have access to a HPC or cloud computing we recommend submitting this as
a batch job as shown in the STAARpipeline tutorial.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="do">### Gene-Centric Coding Analysis</span></span></code></pre></div>
</div>
<div id="burden-scores" class="section level2">
<h2>Burden Scores</h2>
<p>The fourth step of is extracting the burden scores of the significant
rare variant sets. In our manuscript, significance is defined as having
the STAAR-Burden p-value less than <span class="math inline">\(10^{-3}\)</span>, however, an individual can
construct burden scores for any set of rare variant sets. To identify
significant rare variant sets, we first compile the results from the
gene-centric coding analysis.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="do">### Compile Gene-Centric Coding Analysis</span></span></code></pre></div>
<p>Again this code is written for this vignette, when using a HPC or
cloud computing refer to the STAARpipeline tutorial to optimally compile
the results. Once compiled the structure of the results should look
something like this:</p>
</div>
<div id="penalized-regression" class="section level2">
<h2>Penalized Regression</h2>
</div>
<div id="superlearner" class="section level2">
<h2>SuperLearner</h2>
</div>
</div>
<div id="rice-cv" class="section level1">
<h1>RICE-CV</h1>
<div id="superlearner-1" class="section level2">
<h2>SuperLearner</h2>
</div>
</div>
<div id="sessioninfo" class="section level1">
<h1>SessionInfo</h1>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="co">#&gt; R version 4.4.1 (2024-06-14)</span></span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a><span class="co">#&gt; Platform: aarch64-apple-darwin20</span></span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a><span class="co">#&gt; Running under: macOS Sonoma 14.7</span></span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a><span class="co">#&gt; Matrix products: default</span></span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib </span></span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0</span></span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a><span class="co">#&gt; locale:</span></span>
<span id="cb5-11"><a href="#cb5-11" tabindex="-1"></a><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span id="cb5-12"><a href="#cb5-12" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb5-13"><a href="#cb5-13" tabindex="-1"></a><span class="co">#&gt; time zone: America/New_York</span></span>
<span id="cb5-14"><a href="#cb5-14" tabindex="-1"></a><span class="co">#&gt; tzcode source: internal</span></span>
<span id="cb5-15"><a href="#cb5-15" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb5-16"><a href="#cb5-16" tabindex="-1"></a><span class="co">#&gt; attached base packages:</span></span>
<span id="cb5-17"><a href="#cb5-17" tabindex="-1"></a><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span id="cb5-18"><a href="#cb5-18" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb5-19"><a href="#cb5-19" tabindex="-1"></a><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span id="cb5-20"><a href="#cb5-20" tabindex="-1"></a><span class="co">#&gt;  [1] digest_0.6.37     R6_2.5.1          fastmap_1.2.0     xfun_0.47        </span></span>
<span id="cb5-21"><a href="#cb5-21" tabindex="-1"></a><span class="co">#&gt;  [5] cachem_1.1.0      knitr_1.48        htmltools_0.5.8.1 rmarkdown_2.28   </span></span>
<span id="cb5-22"><a href="#cb5-22" tabindex="-1"></a><span class="co">#&gt;  [9] lifecycle_1.0.4   cli_3.6.3         sass_0.4.9        jquerylib_0.1.4  </span></span>
<span id="cb5-23"><a href="#cb5-23" tabindex="-1"></a><span class="co">#&gt; [13] compiler_4.4.1    rstudioapi_0.16.0 tools_4.4.1       evaluate_1.0.0   </span></span>
<span id="cb5-24"><a href="#cb5-24" tabindex="-1"></a><span class="co">#&gt; [17] bslib_0.8.0       yaml_2.3.10       rlang_1.1.4       jsonlite_1.8.9</span></span></code></pre></div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
